## ЗАДАНИЕ ДЛЯ ПРОГРАММИРУЮЩИХ

**Ссылка**, на источник текста

In [ ]:
DATA_URL = "http://az.lib.ru/t/tolstoj_a_k/text_0180.shtml"

Устанавливаем библиотеки

In [ ]:
! pip install -q PyYaml==5.3.1
! pip install -q rnnmorph==0.4.0
! pip install -q nltk==3.2.5
! pip install -q keras==2.1.4
! pip install -q h5py==2.7.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 18.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 56.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 16.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.11.0 requires keras<2.12,>=2.11.0, but you have keras 2.1.4 which is incompatible.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.11.0 requires h5py>=2.9.0, but you have h5py 2.7.0 which is incompatible.
tensorflow 2.11.0 requires keras<2.12,>=2.11.0, but you have keras 2.1.4 which is incompatible.


In [ ]:
%tensorflow_version 1.x
import nltk
import warnings
import urllib.request
from tqdm import tqdm
from nltk import FreqDist
from bs4 import BeautifulSoup
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
warnings.filterwarnings('ignore')
nltk.download('punkt', quiet=True)
nltk.download("stopwords", quiet=True)

from rnnmorph.predictor import RNNMorphPredictor
predictor = RNNMorphPredictor(language="ru")
opener = urllib.request.URLopener({})
resource = opener.open(DATA_URL)
raw_text = resource.read().decode(resource.headers.get_content_charset())

soup = BeautifulSoup(raw_text, features="html.parser")

for script in soup(["script", "style"]):
    script.extract()

cleaned_text = soup.get_text()
tokenized_sentences = [word_tokenize(sentence) for sentence in sent_tokenize(cleaned_text)]
predictions = [[pred.normal_form for pred in sent if pred.normal_form.isalpha()] for sent in tqdm(predictor.predict_sentences(sentences=tokenized_sentences), "sentences") ] # ИСПРАВИЛ НА ТОЛЬКО БУКВЫ
non_uniq_tokens = [word for sentence in predictions for word in sentence]

answers = {}
for i in non_uniq_tokens:
  if i in answers:
    answers[i] += 1
  else:
    answers[i] = 1

STOPWORDS = set(stopwords.words("russian"))

t = 1
counter = more_words_count + 1
while counter > more_words_count:
  counter = 0
  n_answers = []
  for i in answers:
    if (answers[i] > t):
      n_answers.append(i)
      counter += 1
  t += 1

counter = 0
for i in n_answers:
  if i in stopwords.words("russian"):
    counter += 1

counter1 = 0
for i in answers:
  if (answers[i] > tokens_more_count):
    counter1 += 1
print("\n")

print("Введите количество предложений")
print("Ответ:", len(predictions))

print("Введите количество токенов, состоящих только из букв")
print("Ответ:", len(non_uniq_tokens))

print(f"Какую долю среди {more_words_count} самых частотных токенов в произведении занимают слова, не входящие в стоп-лист?")
print("Ответ:", (more_words_count - counter) / more_words_count)

print(f"Сколько токенов встречается в тексте строго больше {tokens_more_count} раз?")
print("Ответ:", counter1)

ValueError: ignored

Создаём объект морфологического анализатора `RNNMorph`

Как видно, текст содержит html теги, от которых нужно избавиться. Выбрасываем из текста HTML-теги с помощью библиотеки Beatiful soap

In [ ]:
from nltk.tokenize import sent_tokenize, word_tokenize
import nltk
nltk.download('punkt')

tokenized_sentences = [word_tokenize(sentence) for sentence in sent_tokenize(cleaned_text)]
"A total of %d 'sentences'" % len(tokenized_sentences)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


"A total of 577 'sentences'"

## Задание 1
С помощью метода `str.isalpha` из стандартной библиотеки Python модифицируйте нижеследующий код так, чтобы в predictions остались только буквенные токены.

In [ ]:
from tqdm import tqdm
predictions = [[pred.normal_form for pred in sent] 
               for sent in tqdm(predictor.predict_sentences(sentences=tokenized_sentences), "sentences") ]
predictions[-11:-10] #Сейчас видно, что токены типа "точка", "запятая" и тд пока присутствуют в предложениях. От них нужно избавиться



sentences: 100%|██████████| 577/577 [00:00<00:00, 157282.99it/s]


[['не',
  'помнить',
  ',',
  'что',
  'произойти',
  'ещё',
  ',',
  'но',
  'когда',
  'я',
  'прийти',
  'в',
  'себя',
  ',',
  'быть',
  'уже',
  'вполне',
  'светлый',
  ',',
  'я',
  'лежать',
  'на',
  'дорога',
  ',',
  'а',
  'рядом',
  'издыхать',
  'мой',
  'конь',
  '.']]

Проверьте себя. Должны получиться следующие значения:

*   Предложений: 577 (возможны расхождения в несколько предложений)
*   Токенов: примерно 8621 (возможны расхождения в некоторое количество токенов)

In [ ]:
len(predictions)

577

In [ ]:
non_uniq_tokens = [word for sentence in predictions for word in sentence]
len(non_uniq_tokens) 


8621

Для продолжения работы над заданием числа должны быть близки к указанным

## Задание 2

Используя `non_uniq_tokens`, стоп-слова для русского языка из библиотеки nltk (`nltk.corpus.stopwords`) и `nltk.FreqDist`, вычислите, **какую долю среди 100 самых частотных** токенов в произведении занимают токены, **не относящиеся** к стоп словам. 

**Например**, если среди 100 самых частотных слов встречается 25 слов, входящих в стоп лист, значит не входят в стоп лист 75 слов, и их доля составит 0.75. 

**Не бойтесь использовать документацию NLTK и тьюториалы.**

In [ ]:
import nltk
from nltk import FreqDist
from nltk.corpus import stopwords
nltk.download("stopwords")
STOPWORDS = set(stopwords.words("russian"))
stopwords.words("russian")[:5] #Пример стоп слов

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


['и', 'в', 'во', 'не', 'что']

Проверьте себя: должно получиться 0.49 (допустимо небольшое расхождение)

## Задание 3
Вычислите, сколько токенов встречается в тексте **строго больше** 50 раз.

Проверьте себя: должно получиться значение 22 (возможно небольшое расхождение)
